src/irmin/store_intf.ml

```diff
@@ -377,25 +394,30 @@ module type S = sig
          and type key := key
          and type metadata := metadata
          and type contents := contents
+         and type contents_key := contents_key
          and type node := node
          and type hash := hash
 
     (** {1 Import/Export} *)
 
-    val hash : tree -> hash
-    (** [hash r c] it [c]'s hash in the repository [r]. *)
-
-    type kinded_hash := [ `Contents of hash * metadata | `Node of hash ]
+    type kinded_key =
+      [ `Contents of contents_key * metadata | `Node of node_key ]
+    [@@deriving irmin]
     (** Hashes in the Irmin store are tagged with the type of the value they
         reference (either {!contents} or {!node}). In the [contents] case, the
         hash is paired with corresponding {!metadata}. *)
 
-    val of_hash : Repo.t -> kinded_hash -> tree option Lwt.t
```

Comment:
As `Store.Commit.of_hash` remains, I think we should keep a `Store.Tree.of_hash` too for simplicity of use.

Reply:
Good idea. I've added back both `Store.Tree.of_hash` and `Store.Contents.of_hash`.